[Xiong Jingya] iP

xiongjya opened the pull request:
# Bubbles
> "You should do what you want to do!" - Bubbles, Powerpuff Girls

Bubbles, the cutest Powerpuff Girl, helps you to track what you have to do! (And you can add ANY TASK into her pocket list!) It's:
* text-based
* easy to learn
* SUPER encouraging <3

All you need to do is, 
1. download it from [here](https://github.com/xiongjya/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. and let it manage your tasks for you 🧚🏻‍♀️

## Features
- task management (different types of tasks such as to-dos, deadlines and events)
- eg.
  - [x] Beat The Amoeba Boys
  - [ ] Beat Mojo Jojo
- managing deadlines (coming soon!)
- reminders (coming soon!)

Commit:
Add Gradle support

xiongjya committed:
Create skeletal version of Bubbles, that greets, echos and exits

xiongjya committed:
Add the ability to store and display text entered by user

xiongjya committed:
Add extra text for more personality of the chatbot

xiongjya committed:
Change previous Task class to ToDoList class, shift features around

xiongjya committed:
Add mark as done feature to each task

xiongjya committed:
Separate tasks to todo, deadline, event tasks

xiongjya committed:
Add input/output redirection technique to semi-automate testing of bot

xiongjya committed:
Separate java classes by their functionality, add error handling

xiongjya committed:
Change data storage of tasks from array[] to Java ArrayList

xiongjya committed:
Add support for deleting tasks from the list

xiongjya committed:
Use enum instead of using if-else branches in taskListener

xiongjya committed:
Save tasks in the hard disk automatically when there are changes

xiongjya committed:
Debug reading due/event dates for Deadline and Event from file

xiongjya committed:
Add functionality for Bubbles to understand date input

xiongjya committed:
Merge branch 'branch-level-7'

xiongjya committed:
Merge branch 'branch-level-8'

xiongjya committed:
Makee the code more OOP

xiongjya committed:
Organize the classes into suitable java packages - bubbles

xiongjya committed:
Organize tasks package to be under bubbles package

xiongjya committed:
Add JUnit tests to test the behavior of the code

xiongjya committed:
Package the app as an executable JAR file for easy distribution

xiongjya committed:
Add JavaDoc comments to the code (public methods)

xiongjya committed:
Tweak the code to comply with a given coding standard

xiongjya committed:
Merge branch 'branch-A-Javadoc'

xiongjya committed:
Merge branch 'branch-A-CodingStandard'

xiongjya committed:
Give users a way to find a task by searching for a keyword

xiongjya committed:
Merge branch 'branch-Level-9'

xiongjya committed:
Merge branch 'add-gradle-support'

xiongjya committed:
Set up Gradle to allow automation of project builds using Gradle

eeliana commented on src/main/java/bubbles/Storage.java:
```diff
@@ -0,0 +1,109 @@
+        while (sc.hasNext()) {
+            String task = sc.nextLine();
+            Object[] inputs = parser.formatTask(task);
+
+            taskList.addTask((String) inputs[0], (String) inputs[1], (Boolean) inputs[2]);
```
Perhaps a comment here specifying what are inputs[0], inputs[1], inputs[2] would be good? It'd be easier to refer to since Parser is in a separate file.

eeliana commented on src/main/java/bubbles/tasks/Deadline.java:
```diff
@@ -0,0 +1,69 @@
+    public String format() {
+        String format = "D | ";
+
+        if (this.isDone) {
+            format += "1 | ";
+        } else {
+            format += "0 | ";
+        }
```
What do you think about having Task do this (format + super.format() + ...)? Since this seems repeated across the 3 types of task and the format() method in Task seems to have the same functionality as its toString() method.

eeliana reviewed (commented):
LGTM. Naming is clear and code is easy to read and understand 👍

xiongjya committed:
Add a GUI to Bubbles using JavaFX technology

xiongjya committed:
Merge branch 'branch-Level-10'

SajalVaishnav commented on src/main/java/bubbles/Bubbles.java:
```diff
@@ -0,0 +1,55 @@
+        ui.echoTask(this.storage);
+    }
+
+    private static void formatting(String str) {
```
Great idea to keep the output neat!

SajalVaishnav commented on src/main/java/bubbles/tasks/Task.java:
```diff
@@ -0,0 +1,75 @@
+     * @param dateInput Date input form the user, formatted in yyyy-mm-dd
+     * @return LocalDate Object representing the input date
+     */
+    public static LocalDate formatDate(String dateInput) {
```
Maybe this does not need to be handled here since not all tasks need to handle dates?

SajalVaishnav commented on src/test/java/bubbles/ParserTest.java:
```diff
@@ -0,0 +1,22 @@
+
+public class ParserTest {
+    @Test
+    public void testFormatTask() {
```
Maybe indicating the input in the method name here would get closer to the module's code requirements?

SajalVaishnav commented on src/test/java/bubbles/tasks/DeadlineTest.java:
```diff
@@ -0,0 +1,20 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class DeadlineTest {
```
Great to see extensive testing beyond the requirements!

SajalVaishnav reviewed (commented):
LGTM! very well written code!

Morrow1ndy commented on src/main/java/bubbles/Bubbles.java:
```diff
@@ -0,0 +1,55 @@
+
+        String introduction = "You should do what you want to do!\n"
+                + "Hello I'm Bubbles from the Powerpuff Girls, what are you up to?";
+        System.out.println("Hello from\n" + logo);
```
Maybe it would better to shift the UI related output message into ui package?

xiongjya replied:
Thank you for your comment! I had created a new Messages class that contains the common messages that would be used by Bubbles.

Morrow1ndy commented on src/main/java/bubbles/Parser.java:
```diff
@@ -0,0 +1,45 @@
+     *          (and date), and the last argument refers to whether the task had been
+     *          completed
+     */
+    public Object[] formatTask(String task) {
```
Perhaps using another data structure other than an array of Objects would be better for self reference and debugging purposes?

Morrow1ndy commented on src/main/java/bubbles/Storage.java:
```diff
@@ -0,0 +1,109 @@
+     * the parser that decodes the user's inputs, and a task list for the Bubbles bot.
+     */
+    public Storage() {
+        this.parser = new Parser();
```
Maybe removing `this` to make coder cleaner?

Morrow1ndy commented on src/main/java/bubbles/Ui.java:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    private static void formatting(String str) {
```
Perhaps the naming of the method here violates the naming convention?

xiongjya replied:
I see, and I do agree with you. I have removed this method.

Morrow1ndy commented on src/main/java/bubbles/tasks/Task.java:
```diff
@@ -0,0 +1,75 @@
+     * @param dateInput Date input form the user, formatted in yyyy-mm-dd
+     * @return LocalDate Object representing the input date
+     */
+    public static LocalDate formatDate(String dateInput) {
+        // assume that the acceptable format for dates related to deadline and event are in the form of
+        // "year-month-day"
+        LocalDate ld = LocalDate.parse(dateInput);
+
+        return ld;
```
Maybe this logic can be shifted to TaskList?

Morrow1ndy commented on src/main/java/bubbles/tasks/ToDo.java:
```diff
@@ -0,0 +1,65 @@
+     * @return The created ToDo Object
+     * @throws EmptyTaskException Exception thrown when the ToDo description is empty
+     */
+    public static ToDo addToDo(String input, boolean isDone) throws EmptyTaskException {
+        if (input.equals("")) {
+            throw new EmptyTaskException("todo");
+        }
+
+        ToDo item = new ToDo(input, isDone);
+
+        return item;
+    }
```
Maybe this logic can be shifted to  TaskList as well?

Morrow1ndy reviewed (commented):
Overall your code is very easy to read!

xiongjya committed:
Edit code to comply with coding style violations after using checkStyle

xiongjya committed:
Create a new Bubbles JAR file

xiongjya committed:
Change JavaFX API version to resolve compile error warnings

xiongjya committed:
Implement assert feature to document important assumptions

The program state needs to be checked at certain points to ensure
that the code is behaving as expected.

For example, assert is used in the TaskList class, to check whether the
count number that tracks the number of tasks in the tasklist is
properly decremented (when a task is deleted) or incremented (when a
new task is added).

Let's include assertions in all of the methods in TaskList, to ensure
that certain states of the task list are right and error-free.

An AssertionError with details will be thrown if the program state is
not as expected.

Assertions are used to provide more details on what went wrong when
errors are thrown (, that are visible to the author of the code).

xiongjya committed:
Merge pull request #1 from xiongjya/branch-A-Assertions

Implement Java assert to document important assumptions

xiongjya committed:
Refactor to improve the code quality where necessary

Code quality is needed to to guarantee understandability and
readability. This is especially crucial if the code might be passed on
to other developers, or if the code is posted to public platforms such
as Github.

For example, to reduce code duplicate, let's extract the formatting of
messages within each method in the TaskList class, and create a new
method in the Message class to deal with the functionality instead.

Furthermore, messages are extracted to be part of the Message class,
on top of the previously implemented messages.

Some methods may be longer than 30 lines of code, but it is due to the
inclusion of white space and assert.

The unsafe practices mentioned in the Code Quality guide are avoided.

xiongjya committed:
Merge pull request #2 from xiongjya/branch-A-CodeQuality

Refactor to improve the code quality where necessary

xiongjya committed:
Use Java Streams for find command

Using Streams improves on the code readability, and cuts down the
number of lines within the method for the find operation.

Let's use Streams, with the filter() and collect() methods instead
of the previous for loop to iterate through the task list.

xiongjya committed:
Merge pull request #3 from xiongjya/branch-A-Streams

Use Java Streams for find command

xiongjya committed:
Add a new remind command

A new feature that reminds the user about the upcoming deadlines and
events, when the user types 'remind' in the command line of the
Bubbles bot. This feature improves on the functionality of the
task tracker bot and thus is able to help users track tasks
in a more efficient manner.

Let's generate a list of upcoming deadlines and events (happening
within a week from the time of command) and list them out to the
user.

Additional information that could be included in the remind message
includes the overdue deadlines and events, that might be included
in the next commit.

xiongjya committed:
Remove use of formatting line for messages replied by the bot

xiongjya committed:
Improve GUI and set up CI with Java and Gradle

xiongjya committed:
Improve the GUI to make it more polished.

Modification includes a change in the font for messages from
Bubbles, and change of display for the user 'Send' button.
An improvement to the help command is made additionally.

xiongjya committed:
Set theme jekyll-theme-cayman

xiongjya committed:
Set theme jekyll-theme-cayman

xiongjya committed:
Add a User Guide to the project

xiongjya committed:
Merge branch 'master' of https://github.com/xiongjya/ip

xiongjya committed:
Improve the User Guide

xiongjya committed:
Improve the User Guide

xiongjya committed:
Improve the User Guide

xiongjya committed:
Improve error handling for invalid/unsupported commands

xiongjya committed:
Edit the code to comply with checkstyle warnings

xiongjya committed:
Edit Continuous Integration (CI) to exclude IO

xiongjya committed:
Change build.gradle file to make the .jar file executable on other OS

xiongjya committed:
Increase reminder date range from within the next 6 days to 7 days

xiongjya committed:
Abstract out reply messages into the Message class
